src/test/denialofservice_tests.cpp

```diff
@@ -356,16 +358,39 @@ BOOST_AUTO_TEST_CASE(DoS_bantime)
     peerLogic->FinalizeNode(dummyNode.GetId(), dummy);
 }
 
-static CTransactionRef RandomOrphan()
+static CTransactionRef OrphanByIndex(const uint256& orphan_index)
 {
     std::map<uint256, COrphanTx>::iterator it;
     LOCK2(cs_main, g_cs_orphans);
-    it = mapOrphanTransactions.lower_bound(InsecureRand256());
+    it = mapOrphanTransactions.lower_bound(orphan_index);
     if (it == mapOrphanTransactions.end())
         it = mapOrphanTransactions.begin();
     return it->second.tx;
 }
 
+static CTransactionRef RandomOrphan()
+{
+    return OrphanByIndex(InsecureRand256());
+}
+
+/** This function runs CPubkey::Verify with parameters that result in all branches of the function ecdsa_signature_parse_der_lax to run. Namely, the signature literal
+ * in vch_sig_str in the following function is comprised of R and S values which both contain leading zeroes, forcing some branches of said function to run which
+ * would otherwise not. This function is called at the end of the DoS_mapOrphans test to force deterministic coverage.
+ */
+static void ForceCoverageInPubKeyVerify()
+{
+    std::string coverage_pubkey_str("\x03\x1c\xef\xd3\xfa\x1e\x91\xd2\x24\x7e\x9d\x74\xa1\xf4\x31\x27\x6e\xe4\x74"
+        "\x5b\xaf\x73\x1d\xb0\x0c\x07\xa0\x78\x81\xa5\xca\x4c\xc9", 33);
+    CPubKey coverage_pubkey(coverage_pubkey_str.begin(), coverage_pubkey_str.end());
+    std::string vch_sig_str("\x30\x44\x02\x20\x00\xa8\x9f\x92\xf4\x47\x6e\x3f\x0b\x1f\x58\x9f\x6b\x3b\xb9\xae\xc0"
+        "\x99\x84\x82\x22\x40\x68\xf5\x12\xf7\x43\xbe\xa7\x54\x87\x7c\x02\x20\x00\xa1\x09\x98\x60\x32\x32\x71\x34"
+        "\xad\xf6\x6a\x9c\x2f\xd9\xd4\xf3\xcc\xf4\xc8\x4c\x38\xbb\xd0\xac\xde\xa7\x3d\x66\x28\xa0\xe0", 70);
+    uint256 hash;
+    hash.SetHex(std::string("6af516422fef8a745aff6acdcc84076c77fb2ecd72bd5711df301230ac58fdd5"));
+
+    coverage_pubkey.Verify(hash, std::vector<unsigned char>(vch_sig_str.begin(), vch_sig_str.end()));
+}
```

Comment:
Looks like the wrong location for this. I fail to understand what CPubKey has to do with the orphan map or even DoS?!

Reply:
Thanks for the review! While trying to understand the non-determinism in the coverage, I used the script `contrib/devtools/test_deterministic_coverage.sh`, which uses `gcovr` to show the diff in which lines were and which ones were not executed between two runs with different coverage. 

Over multiple runs of the script, the only two places where the coverage was non-deterministic were: 

- the function `RandomOrphan` in denialofservice_tests.cpp 
- lines 135-136, 147-148 in pubkey.cpp, in the function `ecdsa_signature_parse_der_lax`. 

The latter function is called implicitly from the test `DoS_mapOrphans` via `SignSignature` (`SignSignature` => `ProduceSignature` => `VerifyScript` => `EvalScript` => `CheckSig` => `VerifySignature` => `Verify` => `ecdsa_signature_parse_der_lax`). The private key which generates the signatures in the test `DoS_mapOrphans` is randomly generated, so the `R` and `S` values of the ECDSA signatures for the inputs of the dependant orphans most of the time do not have leading zeros. The lines 135-136, 147-148 run only when there are leading zeroes in `R`, `S`, respectively. This results in those lines sometimes running and sometimes not. 

My proposed solution is to create a constant signature with leading zeroes in both `R` and `S` and force the function `ecdsa_signature_parse_der_lax` to run on that signature by calling `CPubkey::Verify` on that signature and a corresponding pubkey and hash. This forces the lines 135-136, 147-148 in pubkey.cpp to always run at least once, resulting in deterministic test coverage.

Reply:
I see. A smaller alternative would be to set the key to a constant?

Reply:
That could be nice actually if it's acceptable. I didn't want to do it initially because I didn't want to modify the test too much and restricted myself to 'parasitic' changes.
On the other hand - if I'm not mistaken, this will not work directly, since, if we have transaction B which is using as an input an output from transaction A, part of the data used to calculate the hash of that input for B (this hash is then signed to produce `R` and `S`) is the hash of the transaction A, which contains a random element (the initial non-dependant 50 orphans are created with `tx.vin[0].prevout.hash = InsecureRand256()`), so `R` and `S` would again be random.

I guess I could use an engineered, non-random value for `tx.vin[0].prevout.hash` for one of the orphans along with an engineered fixed key. But then again, this test would loose the fuzzing feature of making a different key every run.

What do you think?

Reply:
`InsecureRand256` is an alias for `g_insecure_rand_ctx.rand256()`, so it can be made deterministic with a call to `SeedInsecureRand(/* deterministic */ true)`.

> this test would loose the fuzzing feature of making a different key every run.

This "fuzzing feature" is useless, as a failure could not be reproduced (when it occurs) and thus not debugged, nor fixed.

When fuzzing is desired, the seed would have to be written to a file or stdout.

Reply:
I see. Then I can just set a seed which guarantees the widest coverage in the beginning of the test and that should take care of the non-determinism in both `RandomOrphan` and `ecdsa_signature_parse_der_lax`, making the code much simpler.

Still, I would argue that it's best to leave the fuzzing and write the seed to a file. That would result in an a better coverage over a long period of time. There might be some branches that run very rarely and setting the seed to a constant would prevent them from ever running in the future in the context of this test, reducing it's long-term coverage. As an example the conditions on lines 151, 157, 160 in pubkey.cpp, in `ecdsa_signature_parse_der_lax` may evaluate to false or true. Although their evaluation was deterministic throughout the 1000+ runs I tried and I haven't gone in-depth to try and understand whether they can at all run under this test's conditions or whether they run extremely rarely, but the latter option can be true for other branches I haven't checked.
Unless handling artifacts created by tests involves a lot of overhead, the only disadvantage I see to leaving the fuzzing is more lines of code in the test, some of it being irrelevant to the functionality of the test itself, but I think that is taken care of by a detailed comment explaining the logic for the extra code.

Reply:
Yeah, up to you if you feel strongly. My opinion is that we shouldn't mix unit tests with fuzz tests, which are designed and compiled separately, see https://github.com/bitcoin/bitcoin/blob/master/doc/fuzzing.md

Reply:
I agree that fuzz tests should be separate and I wasn't familiar with `src/test/fuzz`. Thanks for pointing me there. However, reading through https://github.com/bitcoin/bitcoin/pull/16320, there are some comments about the positive value of non-determinism in tests (I was trying to understand if the direction is to completely eliminate non-determinism). I think that an effort should be made when writing new tests towards determinism and separating the fuzzy parts into the fuzz framework, but since this test already has non-determinism, I think I will stick with writing the seed to stdout. Separating the fuzzy from the non-fuzzy parts should perhaps be a part of another pull request.
That brings the question - there are a lot of tests that use random values. I completely agree that randomness in tests is practically useless unless the seed is known. What do you think about setting and writing the seed to stdout as part of the `BasicTestingSetup` fixture so that seeds for all tests are known? This would also prevent code duplication when multiple tests require this.

Reply:
We already use `BOOST_TEST_RANDOM` https://www.boost.org/doc/libs/1_71_0/libs/test/doc/html/boost_test/utf_reference/rt_param_reference/random.html, which is logged. So might as well use that as the seed for our test rng?

Reply:
Sounds good. Do you suggest seeding the rng with `BOOST_TEST_RANDOM` in `BasicTestingSetup` or only in this test for now? I'm thinking the latter, since otherwise it changes the behaviour of all tests that use random numbers when running them locally and it would require adjusting the test execution scripts, as now the seed must be provided externally. For example, `test_deterministic_coverage.sh` will always show deterministic coverage when run locally unless we provide a different value of `BOOST_TEST_RANDOM` every time it's run. I feel handling those issues would be out of the scope of this pull request and may be done separately?

Reply:
> test_deterministic_coverage.sh will always show deterministic coverage when run locally unless we provide a different value of BOOST_TEST_RANDOM every time it's run. I feel handling those issues would be out of the scope of this pull request and may be done separately?

I think this is the way the script should work (report non-determinism even when the rng seed is pinned to a constant). As suggested by you this should be done in a separate pull request. Are you interested in working on this as well? Otherwise I will create an issue for others to pick up.

Reply:
Yes, that sounds interesting, since IMO recording the seed for any test that uses random values is important. I'll check whether pinning the seed still results in non determinism between runs and based on that I'll see what can be done about it and about setting the seed to `BOOST_TEST_RANDOM` in `BasicTestingSetup`.
For now, I guess setting the seed to `BOOST_TEST_RANDOM` just inside this test would be enough?